packages/xo-web/src/common/modal.js

```diff
@@ -259,57 +260,106 @@ export const confirm = ({ body, icon = 'alarm', title, strongConfirm }) =>
-
-class FormModal extends BaseComponent {
-  state = {
-    value: this.props.defaultValue,
-  }
-
-  get value () {
-    return this.state.value
-  }
-
-  render () {
-    const { body, formId } = this.props
-    return (
-      <form id={formId} onSubmit={preventDefault}>
-        {cloneElement(body, {
-          value: this.state.value,
-          onChange: this.linkState('value'),
-        })}
-      </form>
-    )
-  }
-}
-
-export const form = ({ body, defaultValue, icon, title, size }) => {
-  const formId = generateRandomId()
-  const buttons = [
-    {
-      btnStyle: 'primary',
-      label: _('formOk'),
-      form: formId,
-    },
-  ]
-  return new Promise((resolve, reject) => {
-    modal(
-      <GenericModal
-        buttons={buttons}
-        icon={icon}
-        reject={reject}
-        resolve={resolve}
-        title={title}
-      >
-        <FormModal body={body} defaultValue={defaultValue} formId={formId} />
-      </GenericModal>,
+let formModalState
+export const form = props =>
+  new Promise((resolve, reject) => {
+    formModalState.props = {
       reject,
-      {
-        bsSize: size,
-      }
-    )
+      resolve,
+      showModal: true,
+      ...props,
+    }
+    disableShortcuts()
   })
-}
+
+const getInitialState = () => ({
+  id: generateRandomId(),
+  localValue: undefined,
+  props: undefined,
+})
+export const FormModal = [
+  provideState({
+    initialState: getInitialState,
+    effects: {
+      initialize () {
+        if (formModalState !== undefined) {
+          throw new Error('FormModal is a singleton!')
+        }
+        formModalState = this.state
+      },
+      finalize: () => {
+        formModalState = undefined
+      },
+      close: () => () => {
+        enableShortcuts()
+        return getInitialState()
+      },
+      cancel ({ close }) {
+        this.state.props.reject()
+        close()
+      },
+      submit ({ close }) {
+        const {
+          props: { resolve },
+          value,
+        } = this.state
+        resolve(value)
+        close()
+      },
+      onChange: (_, localValue) => () => ({
+        localValue,
+      }),
+    },
+    computed: {
+      value: ({ props = {}, localValue = props.defaultValue }) => localValue,
+    },
+  }),
+  injectState,
+  ({ state: { id, props = {}, value }, effects }) => (
+    <ReactModal
+      bsSize={props.size}
+      onHide={effects.cancel}
+      show={props.showModal}
+    >
+      <ReactModal.Header closeButton>
+        <ReactModal.Title>
+          {props.icon !== undefined ? (
+            <span>
+              <Icon icon={props.icon} /> {props.title}
+            </span>
+          ) : (
+            props.title
+          )}
+        </ReactModal.Title>
+      </ReactModal.Header>
+
+      <ReactModal.Body>
+        <form id={id}>
+          {props.body !== undefined &&
+            cloneElement(props.body, {
+              value,
+              onChange: effects.onChange,
+            })}
```

Comment:
Use a computed value for this.

Reply:
Unfortunately that's possible because computeds in Freactal cannot access effects…

Reply:
Can you at least add a comment explaining what we would like to do and the current limitations?

Reply:
Sure :+1: